internal/martian/proxy_conn.go

```diff
@@ -40,19 +40,28 @@ type proxyConn struct {
 	cs     tls.ConnectionState
 }
 
-func newProxyConn(p *Proxy, conn net.Conn) *proxyConn {
+func newProxyConn(p *Proxy, conn net.Conn) (*proxyConn, error) {
 	v := &proxyConn{
 		Proxy: p,
 		brw:   bufio.NewReadWriter(bufio.NewReader(conn), bufio.NewWriter(conn)),
 		conn:  conn,
 	}
 
 	if tconn, ok := conn.(*tls.Conn); ok {
+		ctx := context.Background()
+		if p.TLSHandshakeTimeout > 0 {
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(context.Background(), p.TLSHandshakeTimeout)
+			defer cancel()
+		}
+		if err := tconn.HandshakeContext(ctx); err != nil {
+			return nil, fmt.Errorf("failed to do TLS handshake: %w", err)
+		}
```

Comment:
Perhaps this should be extracted to method so that `newProxyConn` does not return error.
This would give us the opportunity to provide better errors.

Reply:
A separate method to do the handshake has been added.